### PR TITLE
EVG-5375 populate expansions on server, expansions for triggers

### PR DESCRIPTION
**Pull request by john-m-liu**
task that ran with the agent changes: https://evergreen-staging.corp.mongodb.com/task/mci_windows_test_operations_07f8a4cc09e7df6e65f8eddc6ef244b469ca2ae8_18_11_01_19_38_38#/log/T

### CODE REVIEW

File: model/project.go [hunk 1]
```diff
@@ -564,42 +564,98 @@ var (
...
 	expansions.Put("version_id", t.Version)
 	expansions.Put("task_id", t.Id)
 	expansions.Put("task_name", t.DisplayName)
 	expansions.Put("build_id", t.BuildId)
 	expansions.Put("build_variant", t.BuildVariant)
-	expansions.Put("workdir", d.WorkDir)
 	expansions.Put("revision", t.Revision)
 	expansions.Put("project", t.Project)
+
+	expansions.Put("workdir", h.Distro.WorkDir)
```
Comment (bsamek) [inline model/project.go]: I think you can delete this line. I think the `workdir` used to refer to the distro's working directory, but now it refers to the task's working directory. The agent overwrites this expansion in `startTask` with the expected value, so the value set here is never seen, and it's confusing to set it to something it doesn't actually refer to.

File: model/project.go [hunk 2]
```diff
@@ -564,42 +564,98 @@ var (
...
 
-	for _, e := range d.Expansions {
+	if t.TriggerID != "" {
+		expansions.Put("trigger_event_identifier", t.TriggerID)
+		expansions.Put("trigger_event_type", t.TriggerType)
+		expansions.Put("trigger_id", t.TriggerEvent)
+		var upstreamProjectID string
+		if t.TriggerType == ProjectTriggerLevelTask {
+			upstreamTask, err := task.FindOneId(t.TriggerID)
+			if err != nil {
```
Comment (bsamek) [inline model/project.go]: Also check that `upstreamTask` is not nil.

File: model/project.go [hunk 3]
```diff
@@ -564,42 +564,98 @@ var (
...
+				return nil, errors.Wrap(err, "error finding task")
+			}
+			expansions.Put("trigger_status", upstreamTask.Status)
+			expansions.Put("trigger_revision", upstreamTask.Revision)
+			upstreamProjectID = upstreamTask.Project
+		} else if t.TriggerType == ProjectTriggerLevelBuild {
+			upstreamBuild, err := build.FindOneId(t.TriggerID)
+			if err != nil {
```
Comment (bsamek) [inline model/project.go]: Also check that `upstreamBuild` is not nil.

File: model/project.go [hunk 4]
```diff
@@ -564,42 +564,98 @@ var (
...
+
+	for _, e := range h.Distro.Expansions {
 		expansions.Put(e.Key, e.Value)
 	}
+	proj := &Project{}
+	err = LoadProjectInto([]byte(v.Config), t.Project, proj)
+	if err != nil {
+		return nil, errors.Wrap(err, "error unmarshaling project")
+	}
+	bv := proj.FindBuildVariant(t.BuildVariant)
```
Comment (bsamek) [inline model/project.go]: Does this just move parsing the project from the server to the agent?

File: rest/client/methods.go [hunk 5]
```diff
@@ -190,6 +190,32 @@ func (c *communicatorImpl) GetVersion(ctx context.Context, taskData TaskData) (*
...
+	info := requestInfo{
+		method:   get,
+		taskData: &taskData,
+		version:  apiVersion1,
+	}
+	info.setTaskPathSuffix("expansions")
+	resp, err := c.retryRequest(ctx, info, nil)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to get expansions for task %s", taskData.ID)
+		return nil, err
```
Comment (bsamek) [inline rest/client/methods.go]: Nit: Combine this will line above.

File: rest/client/methods.go [hunk 6]
```diff
@@ -190,6 +190,32 @@ func (c *communicatorImpl) GetVersion(ctx context.Context, taskData TaskData) (*
...
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusConflict {
+		return nil, errors.New("conflict; wrong secret")
+	}
+	err = util.ReadJSONInto(resp.Body, &e)
+	if err != nil {
+		err = errors.Wrapf(err, "unable to read project version response for task %s", taskData.ID)
+		return nil, err
```
Comment (bsamek) [inline rest/client/methods.go]: Nit: Combine with line above.

File: model/project.go (hunk 4 above)
Comment (tychoish) [reply model/project.go]: this function has to run on the server as it does database calls.

File: model/project.go (hunk 1 above)
Comment (john-m-liu) [reply model/project.go]: This is actually also called from when we spawn hosts from tasks, but using it there is nonsensical so I removed it

File: model/project.go (hunk 2 above)
Comment (john-m-liu) [reply model/project.go]: done

File: model/project.go (hunk 3 above)
Comment (john-m-liu) [reply model/project.go]: done

File: model/project.go (hunk 4 above)
Comment (john-m-liu) [reply model/project.go]: We're doing it in both places now (here only out of necessity to get the variant config)

File: rest/client/methods.go (hunk 5 above)
Comment (john-m-liu) [reply rest/client/methods.go]: done

File: rest/client/methods.go (hunk 6 above)
Comment (john-m-liu) [reply rest/client/methods.go]: done

**Pull request merged**